### PR TITLE
feature/improve-models Add Generic Model class

### LINKED ISSUE #56: Implementing at least one of the standard .NET interfaces (IEnumerable, IComparable, IComparer, IEquatable, IEnumerator, etc.)

### PR DESCRIPTION
**Pull request by onlyidev**
Updates:
- Added Generic (related with #52 ) Model class that implements IEquatable (closes #56)
- Added LINQ query (related with #55)
- Added named method argument usage (related with #50)
- Added a front page for searching through recipes

### CODE REVIEW

File: Controllers/RecipeController.cs [hunk 1]
```diff
@@ -41,11 +43,12 @@ public RecipeController(AppDatabaseContext db)
     [Authorize]
     public IActionResult UpsertRecipe(RecipeModel recipe) //For upserting we need the full model information (id can be ommited for creating a new recipe)
     {
-        var email = (HttpContext.User.Identity as ClaimsIdentity)?.getEmailFromClaim();
+        recipe.Owner = (HttpContext.User.Identity as ClaimsIdentity)?.getEmailFromClaim() ?? recipe.Owner; // Set recipe owner to current user
 
-        if (recipe.Owner.Equals((HttpContext.User.Identity as ClaimsIdentity)?.getEmailFromClaim()))
+        if ((from r in _db.Recipes
+             where r.Id == recipe.Id
+             select r).AsNoTracking().First()?.Equals(recipe) ?? true) // Check if signatures of the object in db and provided object match (if recipe exists in db)
```
Comment (Storas008) [inline Controllers/RecipeController.cs:50]: Why is this changed? is this a bug fix?

File: Controllers/RecipeController.cs [hunk 2]
```diff
@@ -41,11 +43,12 @@ public RecipeController(AppDatabaseContext db)
     [Authorize]
     public IActionResult UpsertRecipe(RecipeModel recipe) //For upserting we need the full model information (id can be ommited for creating a new recipe)
     {
-        var email = (HttpContext.User.Identity as ClaimsIdentity)?.getEmailFromClaim();
+        recipe.Owner = (HttpContext.User.Identity as ClaimsIdentity)?.getEmailFromClaim() ?? recipe.Owner; // Set recipe owner to current user
```
Comment (Storas008) [inline Controllers/RecipeController.cs:46]: why is it changed to a completely different variable?

Comment (onlyidev) [reply Controllers/RecipeController.cs:46]: This is not a new variable. The flow looks like this:
Receive recipe object from user -> Since recipe can be changed only by the user who owns it - we set the recipes owner to current user -> if this is a create operation we put the recipe into db / if this is an update operation we hard set the provided recipe email to current user and check if the model signature (id + email in this case) of db and provided model match: if they do - we update otherwise we return Bad Rquest

**Review by Azuolas3 — APPROVED** [review]
Aside from minor nitpicks/notes, code looks good
